src/main/tut/2.4-representable-functors.md

```diff
@@ -0,0 +1,223 @@
+      h compose f
+  }
+
+/* type lambda with Kind Projector plugin:
+   new Contravariant[Op[A, ?]] {...} */
+```
+................
+```Haskell
+instance Profunctor (->) where
+  dimap ab cd bc = cd . bc . ab
+  lmap = flip (.)
+  rmap = (.)
+```
+```scala
+implicit def arrowProfunctor =
+  new Profunctor[Function1] {
+    def dimap[A, B, C, D]
+        (ab: A => B)(cd: C => D)(bc: B => C)
+        : A => D =
+      cd compose bc compose ab
+
+    def lmap[A, B, C]
+        (f: C => A)(ab: A => B): C => B =
+      f andThen ab
+
+    def rmap[A, B, C]
+        (f: B => C)(ab: A => B): A => C =
+      f compose ab
+  }
+```
+................
+```Haskell
+C(-, =) :: Cop × C -> Set
+```
+```scala
+val `C(-, =)`: (Cop `×` C) => Set
+```
+................
+```Haskell
+αx :: C(a, x) -> F x
+```
+```scala
+def α[X]: C[A, X] => F[X]
+```
+................
+```Haskell
+alpha :: forall x. (a -> x) -> F x
+```
+```scala
+/*
+In order to make a universal transformation,
+another type needs to be introduced.
+FunctionK (~>) transforms values from one
+first-order-kinded type into another:
+*/
+trait ~>[F[_], G[_]] {
+  def apply[B](fa: F[B]): G[B]
+}
+
+/*
+Kind Projector plugin provides
+a more concise syntax for type lambdas:
+*/
+def alpha[A]: (A => ?) ~> F
+```
+................
+```Haskell
+fmap f . alpha = alpha . fmap f
+```
+```scala
+(fmap(f) _ compose alpha.apply) ==
+  (alpha.apply _ compose fmap(f))
+```
+................
+```Haskell
+fmap f (alpha h) = alpha (f . h)
+```
+```scala
+fmap(f)(alpha(h)) == alpha(f compose h)
+```
+................
+```Haskell
+beta :: forall x. F x -> (a -> x)
+```
+```scala
+def beta[A]: F ~> (A => ?)
+```
+................
+```Haskell
+alpha :: forall x. (Int -> x) -> [x]
+alpha h = map h [12]
+```
+```scala
+def alpha: (Int => ?) ~> List =
+  new ~>[Int => ?, List] {
+    def apply[A](fa: Int => A) =
+      List(12).map(fa)
+  }
+```
+................
+```Haskell
+map f (map h [12]) = map (f . h) [12]
+```
+```scala
+fmap(f)(fmap(h)(List(12))) ==
+  fmap(f compose h)(List(12))
+
+/* Or using scala stdlib:
+List(12).map(h).map(f) ==
+  List(12).map(f compose h) */
+```
+................
+```Haskell
+beta :: forall x. [x] -> (Int -> x)
+```
+```scala
+def beta: List ~> (Int => ?)
+```
+................
+```Haskell
+class Representable f where
+   type Rep f :: *
+   tabulate :: (Rep f -> x) -> f x
+   index    :: f x -> Rep f -> x
+```
+```tut:silent
+trait Representable[F[_]] {
+  type Rep
+
+  def tabulate[X](f: Rep => X): F[X]
+
+  def index[X]: F[X] => Rep => X
+}
+```
+................
+```Haskell
+data Stream x = Cons x (Stream x)
+```
+```tut:silent
+final case class Stream[X](
+    h: () => X,
+    t: () => Stream[X]
+)
+```
+................
+```Haskell
+instance Representable Stream where
+    type Rep Stream = Integer
+    tabulate f = Cons (f 0) (tabulate (f . (+1)))
+    index (Cons b bs) n = if n == 0 then b else index bs (n - 1)
+```
+```tut:silent
+implicit def streamRepresentable =
```

Comment:
I would add the explicit type here, for two reasons:
- it's good practice to always annotate implicits
- it's someone decides to add a signature of `Representable[Stream]` the code will compile, but be broken as the refinement is lost